### PR TITLE
Port to LLVM 16.0.3

### PR DESCRIPTION
**Pull request by ikeycode**
This has been tested in as far as building `moss` and `boulder` against patched LDC with LLVM16.0.3 in a Serpent OS test environment. The binaries function perfectly and perhaps actually a bit smaller than they were before.

### CODE REVIEW

**Comment by ikeycode** [comment]
First PR to LDC so I'm expecting a bunch of test failures =)

**Comment by ikeycode** [comment]
My only "real" concern with this PR is there is no real sane way now to `getDefaultCPU` for AArch64, and I'm not entirely convinced `generic` is a great fallback. Perhaps we could put something better there ?

File: driver/main.cpp [hunk 1]
```diff
@@ -579,7 +579,9 @@ void initializePasses() {
 #endif
   initializeVectorization(Registry);
   initializeInstCombine(Registry);
+#if LDC_LLVM_VER < 1600
   initializeAggressiveInstCombine(Registry);
```
Comment (JohanEngelen) [inline driver/main.cpp:583]: Did you check in LLVM history whether this was renamed or replaced with something else?

**Comment by JohanEngelen** [comment]
> First PR to LDC so I'm expecting a bunch of test failures =)

Thanks!
You should add an LLVM16 testcase in this PR to .github/workflows/supported_llvm_versions.yml.

File: driver/linker.h [hunk 2]
```diff
@@ -42,7 +42,11 @@ bool linkAgainstSharedDefaultLibs();
 /**
  * Returns the -platformlib library names, if specified.
  */
+#if LDC_LLVM_VER < 1600
 llvm::Optional<std::vector<std::string>> getExplicitPlatformLibs();
```
Comment (JohanEngelen) [inline driver/linker.h:46]: I thought that a bunch of these cases could be solved by replacing llvm::Optional with std::optional for all LLVM versions (I think it's remnant from when we did not use C++11?).

**Comment by ikeycode** [comment]
Thanks for the feedback! Working on it now

File: driver/main.cpp (hunk 1 above)
Comment (ikeycode) [reply driver/main.cpp:583]: I also explained this in the commit msg with the refactored PR, but:

 - https://reviews.llvm.org/D137116

**Comment by ikeycode** [comment]
Debugging some cruft with CI missing the llvm-symbolizer..

**Comment by ikeycode** [comment]
Ok so according to the tests @JohanEngelen we can't just switch to std::optional :/

**Pull request closed without merging**